### PR TITLE
[MD] Add design documents of multiple data source feature

### LINKED ISSUE #2524: [MD] Add design docs of MD feature
- [x] create `/doc` folder 
- [x] high level design
- [x] user story
- [ ] Data source management detail design
- [ ] Crypto service detail design

### PR DESCRIPTION
**Pull request by zhongnansu**
Signed-off-by: Su <szhongna@amazon.com>

### Description
Add below design documents 
- high level design
- user stories
- client management design  
 
Add docs to the root folder `./docs/multi-datasource` suggested by @ashwin-pc 

### Issues Resolved
#2524 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by noCharger** [comment]
@zhongnansu  My PR https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2510 contains some significant design updates. I'd appreciate it if we could get it finalized first.

File: src/plugins/data_source/docs/client_management_design.md
Comment (ashwin-pc) [inline src/plugins/data_source/docs/client_management_design.md:7]: I'm not sure if committing this content to code is ideal. Documentation of API's and high level readme's for usage and development make sense, but this seems more suited for either an Issue, wiki or some other medium.

**Review by ashwin-pc — CHANGES_REQUESTED** [review]
I like the idea of sharing design docs, but i dont think this is the right place to add them. Have we considered any other places to share these? Ideally we do that by opening an issue and add it to the contents there.

Developer docs can definitely reside in the code, design docs, i'm not so sure.

cc: @AMoo-Miki @joshuarrrr @kavilla @zengyan-amazon , what are your thought?

File: src/plugins/data_source/docs/client_management_design.md
Comment (seraphjiang) [reply src/plugins/data_source/docs/client_management_design.md:7]: Hi team, just my two cents

design docs should be version controlled like code, build script, testing. It is good to write rfc when project is planning and request for feedback. eventually, we should convert to doc and add to repo. Keep it up to date while architecture evolution

**Comment by zhongnansu** [comment]
> I like the idea of sharing design docs, but i dont think this is the right place to add them. Have we considered any other places to share these? Ideally we do that by opening an issue and add it to the contents there.
> 
> Developer docs can definitely reside in the code, design docs, i'm not so sure.
> 
> cc: @AMoo-Miki @joshuarrrr @kavilla @zengyan-amazon , what are your thought?

@ashwin-pc Glad we on the same page that adding design documents is good practice. 
I prefer adding those doc to repo, for iteration, and better reachability. Another benefit I can think of is external contributors could just use PR of design docs for design review. An [example](https://github.com/opendistro-for-elasticsearch/kibana-reports/pull/33) of reviewing user story of big feature that was contributed from external.

I am open to discussion

**Comment by seraphjiang** [comment]
> @zhongnansu  My PR https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2510 contains some significant design updates. I'd appreciate it if we could get it finalized first.

@noCharger Consider convert the ECD/issue to doc and commit with PR

File: src/plugins/data_source/docs/client_management_design.md
Comment (ashwin-pc) [reply src/plugins/data_source/docs/client_management_design.md:7]: >  eventually, we should convert to doc and add to repo

Good point. Yeah content like this definitely goes a long way in understanding exisiting features. If thats the case we should be more intentional in our wording. e.g. "We are planning to introduce a new `data-source` model" sounds like its still a proposal and not the final design. It should instead read something like "We **_have_** introduced" to indicate that this is the final design and that it is already implemented.

Also I'm not sure if we want to add User Stories to the code. "User Story" are using for software in the development phase, not once its been completely developed and released. 

> User stories are written by or for users or customers to influence the functionality of the system being developed.

**Comment by ashwin-pc** [comment]
Are we setting precedence here or has any other opensearch project done something similar?

**Comment by zhongnansu** [comment]
> Are we setting precedence here or has any other opensearch project done something similar?

Reporting OSD plugin has done the similar. https://github.com/opensearch-project/dashboards-reports/tree/main/docs. Also check my previous comments how external contributors are contributing big feature to reporting, by also updating design docs & user stories.

**Comment by kavilla** [comment]
Couple of comments:

These are only adding the images to source control and not the files that were used to create these images. It would be really hard to recreate these diagrams from scratch as a community member if I need to make an update to those files.

There are references to `OSD`, for functional purposes we got the approval to utilize `OSD`. But references to the application should utilize `OpenSearch Dashboards` if it adds too much clutter then maybe a legend that lets the reader know that OSD = OpenSearch Dashboards. But I'm not sure if that's okay because we've been explicitly told to never reference OpenSearch Dashboards in public documentation as OSD.

There are references to OpenSearch Dashboards as `Dashboards`. Please use OpenSearch Dashboards, dashboards are specific component of OpenSearch Dashboards.

There are references to OpenSearch as `opensearch`. Please make sure this is `OpenSearch`.

Have we verified that this does not get built in to the final distribution?

Thank you so much!

**Comment by zhongnansu** [comment]
@kavilla I have fixed the opensearch/osd naming conventions, thx. For img, I'll update with `puml` files. I also check the min artifacts, it includes the doc folder, same for any other components/plugins, will that be a concern ?
image

File: src/plugins/data_source/docs/user_stories.md
Comment (noCharger) [inline src/plugins/data_source/docs/user_stories.md:57]: We removed credential management for security reason.

File: src/plugins/data_source/docs/user_stories.md
Comment (noCharger) [inline src/plugins/data_source/docs/user_stories.md:69]: OpenSearch 3.0 preview -> OpenSearch 2.4.0 release?

File: src/plugins/data_source/docs/user_stories.md
Comment (noCharger) [inline src/plugins/data_source/docs/user_stories.md:71]: username and password when creating a data source. -> when OpenSearch security plugin is installed?

File: src/plugins/data_source/docs/high_level_design.md
Comment (noCharger) [inline src/plugins/data_source/docs/high_level_design.md:22]: 2.4.0 and also 2.x?

File: src/plugins/data_source/docs/high_level_design.md
Comment (noCharger) [inline src/plugins/data_source/docs/high_level_design.md:141]: And we will also provide the option to integrate with a key store, such as AWS KMS, to use keys with the key store. -> let's put in future plans

**Review by noCharger — COMMENTED** [review]
@zhongnansu  The graphs in client management design are intuitive and thanks for put everything together! I'd appreciate if we can have authors and email in the docs too.

File: src/plugins/data_source/docs/user_stories.md
Comment (zhongnansu) [reply src/plugins/data_source/docs/user_stories.md:57]: Yeah, I'll remove this apart along with the diagram

File: src/plugins/data_source/docs/user_stories.md
Comment (zhongnansu) [reply src/plugins/data_source/docs/user_stories.md:71]: fixed

File: src/plugins/data_source/docs/high_level_design.md
Comment (zhongnansu) [reply src/plugins/data_source/docs/high_level_design.md:141]: I'll remove this part for now, you can add more with details in the credential design doc.

**Comment by ashwin-pc** [comment]
> > I like the idea of sharing design docs, but i dont think this is the right place to add them. Have we considered any other places to share these? Ideally we do that by opening an issue and add it to the contents there.
> > Developer docs can definitely reside in the code, design docs, i'm not so sure.
> > cc: @AMoo-Miki @joshuarrrr @kavilla @zengyan-amazon , what are your thought?
> 
> @ashwin-pc Glad we on the same page that adding design documents is good practice. I prefer adding those doc to repo, for iteration, and better reachability. Another benefit I can think of is external contributors could just use PR of design docs for design review. An [example](https://github.com/opendistro-for-elasticsearch/kibana-reports/pull/33) of reviewing user story of big feature that was contributed from external.
> 
> I am open to discussion

Ah good point @zhongnansu. That is a very valid use-case. My question then becomes, do we want to pull the location of these docs up higher in the repository? like a root level `/docs` folder like the PR you described since these are more architectural docs, they may touch more than a single plugin or component and it may be worth it to colocate them in a single place. E.g. All MDS work is split in 2 plugins, but it also affects the data plugin and how it functions. Changes there cant be harder to track if this doc stays in the `datasource` plugin folder and for more complex designs that touch more plugins

**Comment by zhongnansu** [comment]
@ashwin-pc Does a `/docs/data-source` path from root sound good to you?

**Comment by ashwin-pc** [comment]
@zhongnansu Yep! `./docs/multi-datasource` maybe? Tying the name to a feature rather than the plugin itself. We can use the same folder for all the new projects including Angular migration, Visualization builder and Point in time etc

File: docs/multi-datasource/high_level_design.md
Comment (ashwin-pc) [inline docs/multi-datasource/high_level_design.md:102]: This section is a lot more powerful than it appears to be and i think it deserves a bigger callout. The title is a little misleading too 😛. Based on what I've read so far, **any feature** including visualizations and dashboards should work out of the box with no changes since the multi datasource feature is transparent to the features that rely on the higher level search api and services like saved objects. 

This section currently makes it clear that visualizations and dashboards dont need any change, but that callout is at the end and does not paint a full picture of how nice this design is. Even other features (e.g. Vis builder) works out of the box because of how transparent this feature is to other high level apps

File: docs/multi-datasource/high_level_design.md
Comment (ashwin-pc) [inline docs/multi-datasource/high_level_design.md:121]: 3.0?

File: docs/multi-datasource/user_stories.md
Comment (ashwin-pc) [inline docs/multi-datasource/user_stories.md:28]: again, I this should be the kibana index for accuracy

**Review by ashwin-pc — COMMENTED** [review]
I must say this was a beautiful read! I'm also sorry i jumped the gun a bit by seeing terms like "design docs" and "user stories" which made me incorrectly assume that they were docs created prior to implementation and not a representation of the actual architecture implemented. Note to self on that.

But great job with these docs! I think we should have every project for OSD follow a similar pattern to document large features like this, since its quite useful to understand how larger features work under the hood in detail. I would have killed to have docs like this for some of the core features in OSD. Kudos!

The docs look good and i'm only holding off on approving this to hear your thoughts on some of my comments.

**Comment by seraphjiang** [comment]
@zhongnansu would you take a look @ashwin-pc 's comments?

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md:75]: In this case we still have the server side protection to block any APIs calls to get the data source client

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md:5]: great suggestion, I'll just take it lol

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md:1]: fixed

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md:46]: removed the duplicate link

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md:105]: The doc was written before any implementation happens. Those are meant to be pseudo code. Not real code. But let me refactor this section and some others.

File: docs/multi-datasource/high_level_design.md
Comment (zhongnansu) [reply docs/multi-datasource/high_level_design.md:31]: Yes. The implementation aims to be able to support those. But it's not in the scope of initial launch. I'll reword a bit

File: docs/multi-datasource/high_level_design.md
Comment (zhongnansu) [reply docs/multi-datasource/high_level_design.md:76]: I'd like to keep it, since this adds more detail, and a hint for implementation.

File: docs/multi-datasource/client_management_design.md
Comment (zhongnansu) [reply docs/multi-datasource/client_management_design.md]: updated this section to reflect the latest change

**Comment by zhongnansu** [comment]
@ashwin-pc Thanks for all the comments! I addressed all of them, and also update the implementation section to reflect the latest change. Could you take another look

File: docs/multi-datasource/high_level_design.md
Comment (ashwin-pc) [reply docs/multi-datasource/high_level_design.md:102]: Like I called out earlier, can we update this section to be a little clearer? This feels like an "Integrating saved objects with Multi datasource" kinda section and not visualizations specifically. Would it be possible to describe when and how to integrate with it? That was my takeaway from this section.

What i'm looking for here is a more general summary of the relationship between saved objects and milti datasource. We have a lot more than 3 saved object types and afaik, only index patterns need to be modified to support muti datasources since they are the object that all other objects reference, and the one that need to know which datasource to call.

Comment (zhongnansu) [reply docs/multi-datasource/high_level_design.md]: "data source " itself is also a "saved object". Not every saved object is related to multiple data source feature. For example, ui-settings, dql query, has nothing to do with data source, so "integrating saved objects with multiple datasource" is not the point we try to make here. We are only referencing "data source " to "index-pattern", as a high level mental model to answer the question how to introduce data source to visualization. The referencing itself is a common way how save objects to integrate with each other, besides it's impl detail, and this doc only talks about high level stuff.  

For other saved object type such as vis-builder, if they reference index-pattern, or reference any save object that references index-pattern. Getting data from data source will be supported out of the box (I'll added it to this section)

This section focuses on explaining the change of a common use case after adopting MD feature, which is visualization in OSD. Word visualization here doesn't actually mean the visualization saved object, it's the solution. 

I updated the wording and title a little bit.
@ashwin-pc

**Review by ashwin-pc — APPROVED** [review]

**Review by kristenTian — APPROVED** [review]

**Pull request merged**
